lib/WP_Auth0_LoginManager.php

```diff
@@ -502,71 +504,79 @@ private function do_login( $user, $userinfo, $is_new, $id_token, $access_token )
-		$auto_login = absint( $this->a0_options->get( 'auto_login' ) );
-
-		if ( $slo && isset( $_REQUEST['SLO'] ) && isset( $_REQUEST['redirect_to'] ) ) {
-			wp_redirect( $_REQUEST['redirect_to'] );
+		$is_sso        = (bool) $this->a0_options->get( 'sso' );
+		$is_slo        = (bool) $this->a0_options->get( 'singlelogout' );
+		$is_auto_login = (bool) $this->a0_options->get( 'auto_login' );
+
+		// Redirected here after checkSession in the footer (templates/auth0-singlelogout-handler.php).
+		if ( $is_slo && isset( $_REQUEST['SLO'] ) ) {
+			if ( ! empty( $_REQUEST['redirect_to'] ) && filter_var( $_REQUEST['redirect_to'], FILTER_VALIDATE_URL ) ) {
+				$redirect_url = $_REQUEST['redirect_to'];
+			} else {
+				$redirect_url = home_url();
+			}
+			wp_redirect( $redirect_url );
 			exit;
 		}
 
-		if ( $sso ) {
-			wp_redirect(
-				'https://' . $this->a0_options->get( 'domain' ) .
-				'/v2/logout?federated&returnTo=' . rawurlencode( home_url() ) . '&client_id=' . $client_id .
-				'&auth0Client=' . base64_encode( wp_json_encode( WP_Auth0_Api_Client::get_info_headers() ) )
+		// If SSO is in use, redirect to Auth0 to logout there as well.
+		if ( $is_sso ) {
+			$telemetry_headers = WP_Auth0_Api_Client::get_info_headers();
+			$redirect_url      = sprintf(
+				'https://%s/v2/logout?returnTo=%s&client_id=%s&auth0Client=%s',
+				$this->a0_options->get( 'domain' ),
+				rawurlencode( home_url() ),
+				$this->a0_options->get( 'client_id' ),
+				$telemetry_headers['Auth0-Client']
 			);
+			wp_redirect( $redirect_url );
 			exit;
 		}
 
-		if ( $auto_login ) {
+		// If auto-login is in use, cannot redirect back to login page.
+		if ( $is_auto_login ) {
 			wp_redirect( home_url() );
 			exit;
 		}
 	}
 
 	/**
+	 *
 	 * Outputs JS on wp-login.php to log a user in if an Auth0 session is found.
 	 * Hooked to `login_message` filter.
+	 * IMPORTANT: Internal callback use only, do not call this function directly!
 	 *
 	 * @param string $previous_html - HTML passed into the login_message filter.
 	 *
 	 * @see WP_Auth0_LoginManager::init()
+	 *
+	 * @return mixed
 	 */
 	public function auth0_sso_footer( $previous_html ) {
 
-		// TODO: Filter handling is incorrect here!
-		echo $previous_html;
-
-		if ( is_user_logged_in() ) {
-			return;
+		// No need to checkSession if already logged in.
+		// URL parameter `no_sso` is set to skip checkSession.
+		if ( is_user_logged_in() || ! empty( $_GET['no_sso'] ) || ! $this->a0_options->get( 'sso' ) ) {
+			return $previous_html;
 		}
 
-		$lock_options = new WP_Auth0_Lock10_Options();
-
-		$sso = $lock_options->get_sso();
-
-		if ( $sso ) {
-			$client_id = $lock_options->get_client_id();
-			$domain    = $lock_options->get_domain();
-			$cdn       = $this->a0_options->get( 'auth0js-cdn' );
-
-			wp_enqueue_script( 'wpa0_auth0js', $cdn );
-			include WPA0_PLUGIN_DIR . 'templates/auth0-sso-handler-lock10.php';
-		}
+		wp_enqueue_script( 'wpa0_auth0js', $this->a0_options->get( 'auth0js-cdn' ) );
+		ob_start();
```

Comment:
what's this? Can't find it in this diff. Also the one `ob_get_clean()`

Reply:
`ob_start()` is output buffer, `ob_get_clean()` returns the buffer and clears it out. This is used to "capture" an `echo` basically

Reply:
So you're creating a new buffer and putting in the contents of the `'templates/auth0-sso-handler-lock10.php'` file, and returning `$previous_html` plus the buffer contents?